opacus/tests/privacy_engine_test.py

```diff
@@ -338,47 +340,84 @@ def test_deterministic_run(self):
         Tests that the noise level is correctly set
         """
-        # Initialize models with parameters to zero
-        model, optimizer, dl, _ = self._init_private_training(
-            noise_multiplier=noise_multiplier
-        )
-        for p in model.parameters():
-            p.data.zero_()
 
-        # Do max_steps steps of DP-SGD
-        n_params = sum([p.numel() for p in model.parameters() if p.requires_grad])
-        steps = 0
-        for x, y in dl:
-            optimizer.zero_grad()
-            logits = model(x)
-            loss = logits.view(logits.size(0), -1).sum(dim=1)
-            # Gradient should be 0
-            loss.backward(torch.zeros(logits.size(0)))
+        def helper_test_noise_level(
+            noise_multiplier: float, max_steps: int, secure_mode: bool
+        ):
+            torch.manual_seed(100)
+            # Initialize models with parameters to zero
+            model, optimizer, dl, _ = self._init_private_training(
+                noise_multiplier=noise_multiplier,
+                secure_mode=secure_mode,
+            )
+            for p in model.parameters():
+                p.data.zero_()
 
-            optimizer.step()
-            steps += 1
+            # Do max_steps steps of DP-SGD
+            n_params = sum([p.numel() for p in model.parameters() if p.requires_grad])
+            steps = 0
+            for x, y in dl:
+                optimizer.zero_grad()
+                logits = model(x)
+                loss = logits.view(logits.size(0), -1).sum(dim=1)
+                # Gradient should be 0
+                loss.backward(torch.zeros(logits.size(0)))
 
-            if max_steps and steps >= max_steps:
-                break
+                optimizer.step()
+                steps += 1
+
+                if max_steps and steps >= max_steps:
+                    break
+
+            # Noise should be equal to lr*sigma*sqrt(n_params * steps) / batch_size
+            expected_norm = (
+                steps
+                * n_params
+                * optimizer.noise_multiplier ** 2
+                * self.LR ** 2
+                / (optimizer.expected_batch_size ** 2)
+            )
+            real_norm = sum(
+                [torch.sum(torch.pow(p.data, 2)) for p in model.parameters()]
+            ).item()
+
+            self.assertAlmostEqual(real_norm, expected_norm, delta=0.15 * expected_norm)
 
-        # Noise should be equal to lr*sigma*sqrt(n_params * steps) / batch_size
-        expected_norm = (
-            steps
-            * n_params
-            * optimizer.noise_multiplier ** 2
-            * self.LR ** 2
-            / (optimizer.expected_batch_size ** 2)
+        helper_test_noise_level(
+            noise_multiplier=noise_multiplier,
+            max_steps=max_steps,
+            secure_mode=False,
+        )
+        helper_test_noise_level(
+            noise_multiplier=noise_multiplier,
+            max_steps=max_steps,
+            secure_mode=True,
```

Comment:
nit: nice syntax that can help future debugging:

```
with self.subTest(secure_mode=True):
    helper_test_noise_level(...)

with self.subTest(secure_mode=False):
    helper_test_noise_level(...)
```

Doesn't do much, but provides nice summary what exactly failed

Reply:
Cool! I will include this!